File: mini-nn-java/src/main/java/com/mininn/Activation.java
package com.mininn;

/**
 * Activation functions for neural networks.
 */
public enum Activation {
    LINEAR,
    RELU,
    LEAKY_RELU,
    SIGMOID,
    TANH,
    SOFTMAX;

    public Tensor apply(Tensor x) {
        switch (this) {
            case LINEAR:
                return x.copy();
            case RELU:
                return x.apply(v -> Math.max(0, v));
            case LEAKY_RELU:
                return x.apply(v -> v >= 0 ? v : 0.01 * v);
            case SIGMOID:
                return x.apply(v -> 1.0 / (1.0 + Math.exp(-clip(v))));
            case TANH:
                return x.apply(Math::tanh);
            case SOFTMAX:
                return applySoftmax(x);
            default:
                return x.copy();
        }
    }

    public Tensor derivative(Tensor x, Tensor output) {
        switch (this) {
            case LINEAR:
                return Tensor.ones(x.getRows(), x.getCols());
            case RELU:
                return x.apply(v -> v > 0 ? 1.0 : 0.0);
            case LEAKY_RELU:
                return x.apply(v -> v >= 0 ? 1.0 : 0.01);
            case SIGMOID:
                return output.apply(v -> v * (1.0 - v));
            case TANH:
                return output.apply(v -> 1.0 - v * v);
            case SOFTMAX:
                // Softmax derivative is complex; handled in loss gradient
                return Tensor.ones(x.getRows(), x.getCols());
            default:
                return Tensor.ones(x.getRows(), x.getCols());
        }
    }

    private static Tensor applySoftmax(Tensor x) {
        int rows = x.getRows();
        int cols = x.getCols();
        Tensor result = new Tensor(rows, cols);

        for (int i = 0; i < rows; i++) {
            // Find max for numerical stability
            double max = Double.NEGATIVE_INFINITY;
            for (int j = 0; j < cols; j++) {
                max = Math.max(max, x.get(i, j));
            }

            // Compute exp and sum
            double sum = 0;
            double[] exps = new double[cols];
            for (int j = 0; j < cols; j++) {
                exps[j] = Math.exp(x.get(i, j) - max);
                sum += exps[j];
            }

            // Normalize
            for (int j = 0; j < cols; j++) {
                result.set(i, j, exps[j] / sum);
            }
        }

        return result;
    }

    private static double clip(double v) {
        return Math.max(-500, Math.min(500, v));
    }
}

File: mini-nn-python/mini_nn/activation.py
"""Activation functions for neural networks."""

import numpy as np
from enum import Enum, auto
from .tensor import Tensor


class ActivationType(Enum):
    """Types of activation functions."""
    LINEAR = auto()
    RELU = auto()
    LEAKY_RELU = auto()
    SIGMOID = auto()
    TANH = auto()
    SOFTMAX = auto()


class Activation:
    """Activation function implementations."""

    @staticmethod
    def apply(x: Tensor, activation_type: ActivationType) -> Tensor:
        """Apply activation function to tensor."""
        if activation_type == ActivationType.LINEAR:
            return x.copy()
        elif activation_type == ActivationType.RELU:
            return x.apply(lambda a: np.maximum(0, a))
        elif activation_type == ActivationType.LEAKY_RELU:
            return x.apply(lambda a: np.where(a >= 0, a, 0.01 * a))
        elif activation_type == ActivationType.SIGMOID:
            return x.apply(lambda a: 1.0 / (1.0 + np.exp(-np.clip(a, -500, 500))))
        elif activation_type == ActivationType.TANH:
            return x.apply(np.tanh)
        elif activation_type == ActivationType.SOFTMAX:
            return Activation._softmax(x)
        else:
            return x.copy()

    @staticmethod
    def derivative(x: Tensor, output: Tensor, activation_type: ActivationType) -> Tensor:
        """Compute derivative of activation function."""
        if activation_type == ActivationType.LINEAR:
            return Tensor.ones(x.rows, x.cols)
        elif activation_type == ActivationType.RELU:
            return x.apply(lambda a: (a > 0).astype(float))
        elif activation_type == ActivationType.LEAKY_RELU:
            return x.apply(lambda a: np.where(a >= 0, 1.0, 0.01))
        elif activation_type == ActivationType.SIGMOID:
            return output.apply(lambda a: a * (1.0 - a))
        elif activation_type == ActivationType.TANH:
            return output.apply(lambda a: 1.0 - a * a)
        elif activation_type == ActivationType.SOFTMAX:
            # Softmax derivative is complex; handled in loss gradient
            return Tensor.ones(x.rows, x.cols)
        else:
            return Tensor.ones(x.rows, x.cols)

    @staticmethod
    def _softmax(x: Tensor) -> Tensor:
        """Apply softmax activation."""
        data = x.numpy()
        # Subtract max for numerical stability
        shifted = data - np.max(data, axis=1, keepdims=True)
        exp_vals = np.exp(shifted)
        return Tensor(exp_vals / np.sum(exp_vals, axis=1, keepdims=True))
